include/swift/AST/DiagnosticsSema.def

```diff
@@ -872,6 +872,31 @@ ERROR(need_hermetic_seal_to_import_module,none,
       "current compilation does not have -experimental-hermetic-seal-at-link",
       (Identifier))
 
+ERROR(modularization_issue_moved,Fatal,
+      "module reference to top-level %0 broken by a context change, "
+      "was found in %1 when building '%2', "
+      "but now a candidate is found only in %3. "
+      "Audit related headers or uninstall SDK roots.",
+      (DeclName, Identifier, StringRef, Identifier))
+ERROR(modularization_issue_type_changed,Fatal,
+      "module reference to top-level %0 broken by a context change, "
+      "its type changed since building '%2', "
+      "it was in %1 now found in %3. "
+      "Ensure Swift language version match between modules or uninstall SDK roots.",
+      (DeclName, Identifier, StringRef, Identifier))
+ERROR(modularization_issue_not_found,Fatal,
+      "module reference to top-level %0 broken by a context change, "
+      "module '%2' references %0 in %1, but it cannot be found in this build. "
+      "Audit related headers or uninstall SDK roots.",
+      (DeclName, Identifier, StringRef, Identifier))
```

Comment:
@tshortli @bnbarham I'd be interest in having your opinions on these three main diagnostics. You can also see them in the test.

Do you have suggestions on how to make them more actionable or simply more understandable by project owners?

Reply:
I like the general direction of providing more complete diagnostics for these failures. My main criticism is that I'm not sure the specific calls to action here are applicable. Suggesting uninstallation of roots may not help because presumably the root is needed. Suggesting auditing headers is a bit vague; I'm not sure what I would personally do to follow through on that suggestion.

If the problem is that the `.swiftmodule` is stale but the build system/compiler is unable to detect that it ought to rebuild it, maybe what we should be suggesting is to remove the potentially stale `.swiftmodule`?

Reply:
Suggesting to delete the stale swiftmodule makes perfect sense, it would fix it in most roots case.

I'm currently looking into pointing the diagnostics on the swiftmodule file with the broken reference instead of the current `<unknown>:0:`. By itself it should help to put into doubts the SDK or recently installed roots.

The case of auditing headers may be too general to be useful. I'm mostly thinking about the Xcc issue but describing that in a diagnostic is a challenge. I'm considerong moving the recommended solutions to separate notes, so we could make them longer if it helps.

Reply:
Is deleting the swiftmodule actually always an option? It could be a distributed binary module without library evolution enabled couldn't it?

Reply:
> Is deleting the swiftmodule actually always an option? It could be a distributed binary module without library evolution enabled couldn't it?

Yeah, we should only suggest it when the module is resilient I suppose.

Reply:
I reworked those diagnostics to make them simpler and shorter. They are now reported on the swiftmodule file making it more intuitive to see where is the broken reference.

I removed the hints to solutions and potential problems from this PR. I plan on replacing them with notes and further analysis of the context, i.e. hint that the swiftmodule is outdated only when resilient and distributed.